File: src/tools/read.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { readTool } from './read.js';

vi.mock('fs/promises', () => ({
  default: {
    readFile: vi.fn(),
  },
}));

describe('readTool', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe('execute', () => {
    it('reads file content', async () => {
      const fs = await import('fs/promises');
      (fs.default.readFile as ReturnType<typeof vi.fn>).mockResolvedValue('file content');
      const result = await readTool.execute({ path: 'test.txt' });
      expect(result.output).toBe('file content');
      expect(fs.default.readFile).toHaveBeenCalledWith('test.txt', 'utf-8');
    });

    it('returns error when file not found', async () => {
      const fs = await import('fs/promises');
      (fs.default.readFile as ReturnType<typeof vi.fn>).mockRejectedValue(new Error('ENOENT: no such file'));
      const result = await readTool.execute({ path: 'nonexistent.txt' });
      expect(result.output).toContain('ENOENT');
    });

    it('slices lines with offset and limit', async () => {
      const fs = await import('fs/promises');
      (fs.default.readFile as ReturnType<typeof vi.fn>).mockResolvedValue('line1\nline2\nline3\nline4\nline5');
      const result = await readTool.execute({ path: 'test.txt', offset: 2, limit: 2 });
      expect(result.output).toBe('line2\nline3');
    });

    it('uses 1-indexed offset', async () => {
      const fs = await import('fs/promises');
      (fs.default.readFile as ReturnType<typeof vi.fn>).mockResolvedValue('line1\nline2\nline3');
      const result = await readTool.execute({ path: 'test.txt', offset: 1 });
      expect(result.output).toBe('line1\nline2\nline3');
    });

    it('handles offset without limit', async () => {
      const fs = await import('fs/promises');
      (fs.default.readFile as ReturnType<typeof vi.fn>).mockResolvedValue('line1\nline2\nline3\nline4');
      const result = await readTool.execute({ path: 'test.txt', offset: 3 });
      expect(result.output).toBe('line3\nline4');
    });

    it('handles limit only (no offset)', async () => {
      const fs = await import('fs/promises');
      (fs.default.readFile as ReturnType<typeof vi.fn>).mockResolvedValue('line1\nline2\nline3\nline4\nline5');
      const result = await readTool.execute({ path: 'test.txt', limit: 2 });
      expect(result.output).toBe('line1\nline2');
    });
  });

});

File: internal/tools/read_test.go
package tools

import (
	"context"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestReadTool_Execute(t *testing.T) {
	tmp := t.TempDir()
	path := filepath.Join(tmp, "test.txt")
	os.WriteFile(path, []byte("file content"), 0o644)

	rt := NewReadTool()
	result, err := rt.Execute(context.Background(), map[string]any{"path": path}, ToolContext{})
	if err != nil {
		t.Fatalf("unexpected error: %s", err)
	}
	if !strings.Contains(result.Output, "file content") {
		t.Errorf("expected 'file content' in output, got %q", result.Output)
	}
}

func TestReadTool_FileNotFound(t *testing.T) {
	rt := NewReadTool()
	result, _ := rt.Execute(context.Background(), map[string]any{"path": "/nonexistent/file.txt"}, ToolContext{})
	if result.Output == "" || !strings.Contains(result.Output, "no such file") {
		t.Logf("read error output: %s", result.Output)
	}
}

func TestReadTool_OffsetLimit(t *testing.T) {
	tmp := t.TempDir()
	path := filepath.Join(tmp, "lines.txt")
	os.WriteFile(path, []byte("line1\nline2\nline3\nline4\nline5"), 0o644)

	rt := NewReadTool()
	// offset=2 (start from line2), limit=2 = read "line2\nline3"
	result, _ := rt.Execute(context.Background(), map[string]any{"path": path, "offset": float64(2), "limit": float64(2)}, ToolContext{})
	if !strings.Contains(result.Output, "line2") || !strings.Contains(result.Output, "line3") || strings.Contains(result.Output, "line4") {
		t.Errorf("expected lines 2-3 in output, got %q", result.Output)
	}
}

func TestReadTool_OffsetWithoutLimit(t *testing.T) {
	tmp := t.TempDir()
	path := filepath.Join(tmp, "lines.txt")
	os.WriteFile(path, []byte("line1\nline2\nline3\nline4"), 0o644)

	rt := NewReadTool()
	// offset=3 skips first 2 lines
	result, _ := rt.Execute(context.Background(), map[string]any{"path": path, "offset": float64(3)}, ToolContext{})
	if !strings.Contains(result.Output, "line3") || !strings.Contains(result.Output, "line4") || strings.Contains(result.Output, "line2") {
		t.Errorf("expected lines from offset 3, got %q", result.Output)
	}
}

func TestReadTool_LimitOnly(t *testing.T) {
	tmp := t.TempDir()
	path := filepath.Join(tmp, "lines.txt")
	os.WriteFile(path, []byte("line1\nline2\nline3\nline4\nline5"), 0o644)

	rt := NewReadTool()
	// limit=2 lines = first 2 full lines
	result, _ := rt.Execute(context.Background(), map[string]any{"path": path, "limit": float64(2)}, ToolContext{})
	if !strings.Contains(result.Output, "line1") || !strings.Contains(result.Output, "line2") || strings.Contains(result.Output, "line3") {
		t.Errorf("expected first 2 lines, got %q", result.Output)
	}
}

func TestReadTool_MissingPath(t *testing.T) {
	rt := NewReadTool()
	result, _ := rt.Execute(context.Background(), map[string]any{}, ToolContext{})
	if !strings.Contains(result.Output, "Error") || !strings.Contains(result.Output, "path") {
		t.Errorf("expected path error, got %q", result.Output)
	}
}
